### PR TITLE
Classifier: refactor convertWorkflowToUseSteps

### LINKED ISSUE #1781: Branching workflow previous functionality
**Package** lib-classifier

**Is your feature request related to a problem? Please describe.**
Complex workflows including branching tasks don't go to previous step as expected.

**Describe the solution you'd like**
Clicking Previous goes to previous Step.

**Additional context**
Iteration from #1758, related to #687.

### LINKED ISSUE #1939: HMS NHS feedback - NEXT LINE and NEXT buttons
**Describe the bug**
Within the 'LINE BY LINE workflow' of HMS NHS: The Nautical Health Service the last step of the task instructs that, to move to the next line, 'click NEXT LINE and NEXT'. This isn't possible at the moment - there's no NEXT button. 

Screenshot 2020-12-11 at 15 19 59

### LINKED ISSUE #2063: Classifier: branching workflows break for combo tasks
## Package
lib-classifier

## Describe the bug
Branching single choice answers break workflow task navigation if `answer.next` points to a combo task.

## To Reproduce
1. Go to  https://frontend.preview.zooniverse.org/projects/bogden/scarlets-and-blues/classify/workflow/17076
2. Choose 'Name list'.
3. Choose 'Next' from the next step.
4. Choose 'Yes, there is another name'

Compare with https://www..zooniverse.org/projects/bogden/scarlets-and-blues/classify ('Index' workflow) for the expected behaviour, which is that you should navigate back to the `T0` task to record names.

## Additional context
`answer.next` for 'Yes, there is another name' points to `T0`, which is a combo task and so isn't present in the workflow tasks. The branching logic looks up step keys by trying to find that task key in each step in turn, then returns undefined.
https://github.com/zooniverse/front-end-monorepo/blob/31c0fdb700136dc85f8c390c9f533fabb8e8c594/packages/lib-classifier/src/store/AnnotatedSteps/AnnotatedSteps.js#L78-L89

### PR DESCRIPTION
**Pull request by eatyourgreens**
Move `convertWorkflowToUseSteps` into a helper function that returns `{ steps, tasks }` where `steps` and `tasks` are step and task snapshots that can be applied to a `Workflow` model. Test the generated step snapshots for a variety of task structures, including single choice branching tasks and combo tasks. Expect single choice branching answers to be converted from task keys to step keys. Expect task keys from combo tasks to be converted to the appropriate step key, even for `answer.next`.

The bug fix is the addition of this new section to the end of `convertWorkflowToUseSteps`, which converts single choice answers to use step keys.
https://github.com/zooniverse/front-end-monorepo/blob/88e45307680e5bc80745ba46d7b8c5291185dba1/packages/lib-classifier/src/store/helpers/convertWorkflowToUseSteps/convertWorkflowToUseSteps.js#L103-L112

I've also moved the conversion into the `Workflow` model, so that the order of execution is now:
1. Load workflow from Panoptes
2. convert workflow tasks to steps
3. create Workflow
4. create WorkflowSteps

Test URL:
This is the National Archives workflow which uses branching, recursion and combo tasks. Choose 'Name list' and try to fill out more than one name in order to test this PR.
https://localhost:8080/?project=12268&workflow=17076&env=production&demo=true

The HMS NHS Line-by-Line workflow is also fixed on this branch. Transcribing a line, then choosing 'Next line', should take you back to the first step to enter another line.
https://localhost:8080/?project=11429&workflow=15077&subjectSet=85304&env=production&demo=true

Package:
lib-classifier

Closes #2063.
Closes #1939.
Closes #1781.

# Review Checklist

## General

- [x] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [x] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

**Comment by eatyourgreens** [comment]
This might also fix #1939.

**Review by mcbouslog — APPROVED** [review]
This looks great! I really like the `convertWorkflowToUseSteps.spec.js` tests. A few comments/suggestions, though none blocking.

**Comment by mcbouslog** [comment]
With this and the recent related PRs, I think #1781 can also be closed?

File: packages/lib-classifier/src/store/helpers/convertWorkflowToUseSteps/convertWorkflowToUseSteps.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    if (Object.keys(tasks).includes(step.next)) {
+      step.next = getNextStepFromTaskKey(step.next, steps, tasks)
+    }
+  })
+
+  // convert single choice answers to use step keys
+  Object.values(tasks).forEach(task => {
+    if (task.type === 'single') {
+      task.answers?.forEach(answer => {
+        if (answer.next?.startsWith('T')) {
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/helpers/convertWorkflowToUseSteps/convertWorkflowToUseSteps.js:107]: Good point. There are definitely workflows where the first task key is `init`. I think the project builder enforces the `T0` format.

[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/helpers/convertWorkflowToUseSteps/convertWorkflowToUseSteps.js:107]: I might put this into a little `taskExists(taskKey)` helper.

File: packages/lib-classifier/src/store/WorkflowStepStore/WorkflowStepStore.js [hunk 2]
```diff
@@ -170,115 +160,13 @@ const WorkflowStepStore = types
...
-        const task = workflow.tasks[taskKey]
-        if (task.type !== 'shortcut') {
-          let stepTasks = [taskKey]
-          if (task.type === 'combo') {
-            stepTasks = task.tasks
-          }
-
-          let stepSnapshot = {
-            next: task.next, // temporarily set next to task key, convert to step key once steps created
-            previous: `S${index}`,
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/WorkflowStepStore/WorkflowStepStore.js:229]: It's still needed as an intermediate step, between creating the steps and linking them up (somewhere around line 80 of `convertWorkflowToUseSteps`.) Hopefully, the tests will catch any cases where the task keys aren't replaced by step keys.

**Comment by eatyourgreens** [comment]
I think you're right, this closes #1781 too.

File: packages/lib-classifier/src/store/helpers/convertWorkflowToUseSteps/convertWorkflowToUseSteps.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  })
+
+  // convert step.next from task key to step key
+  steps.forEach(([stepKey, step]) => {
+    if (Object.keys(tasks).includes(step.next)) {
+      step.next = getNextStepFromTaskKey(step.next, steps, tasks)
+    }
+  })
+
+  // convert single choice answers to use step keys
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/helpers/convertWorkflowToUseSteps/convertWorkflowToUseSteps.js:107]: Yep, this is the big change. Hopefully the missing piece to get looping transcription workflows working.

**Pull request merged**